Linux arm64 release

#### Describe your feature request

AArch64 CI build has been disabled in c572bdfb4f0a571c65dc87e601280734387a66bb. It would be nice to bring this back.

As shown in <https://github.com/helix-editor/helix/issues/1647#issuecomment-1164545327>, the following steps are necessary to re-enable the AArch64 builds.
- [x] Support grammar cross-compilation - #3137
- [ ] Fix C++ language version incompatibility in grammars
    - [ ] Bump `tree-sitter-php` to <https://github.com/tree-sitter/tree-sitter-php/commit/866e4a155739a1374da5247b876e70f8639005f6> or later. The current version [fails to build](https://github.com/yvt/helix/runs/6967639987?check_suite_focus=true#step:6:531) because the cross-compilation environment created by cross-rs provides an older version of GCC, which can't compile a version of `tree-sitter-php` prior to <https://github.com/tree-sitter/tree-sitter-php/pull/126>.
    - [x] Bump `tree-sitter-elm` to <https://github.com/elm-tooling/tree-sitter-elm/commit/a9a8efad446f78db3989d7ed8517987daf510c83> or later. Ditto.
- [ ] Or, upgrade the cross-rs environment by waiting for cross-rs 0.3.x (cross-rs/cross#591), [supplying a custom Dockerfile](https://github.com/cross-rs/cross#dockerfiles), or installing cross-rs from source
- [ ] Update [`.github/workflows/release.yml`](https://github.com/helix-editor/helix/blob/master/.github/workflows/release.yml#L45)

@archseer Oh, yay, and just as I’m setting up my development environment on a Pinebook Pro ;)

I just tried building from source on it on postmarketOS (based on Alpine) and it failed with:

```
   Compiling helix-lsp v0.6.0 (/home/user/Projects/other/helix/helix-lsp)
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crt1.o" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crti.o" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtbegin.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.0.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.1.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.10.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.11.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.12.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.2.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.3.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.4.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.5.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.6.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.7.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.8.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.build_script_build.c17cca23-cgu.9.rcgu.o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6.2qmb9eu1yn7c20v0.rcgu.o" "-Wl,--as-needed" "-L" "/home/user/Projects/other/helix/target/release/deps" "-L" "/home/user/Projects/other/helix/target/release/build/tree-sitter-68cce7d026d726dc/out" "-L" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib" "-Wl,-Bstatic" "/home/user/Projects/other/helix/target/release/deps/libhelix_loader-47727e8a9ea74c4d.rlib" "/home/user/Projects/other/helix/target/release/deps/libonce_cell-0eaa9390835d2a8a.rlib" "/home/user/Projects/other/helix/target/release/deps/libcc-cad31f8833909884.rlib" "/home/user/Projects/other/helix/target/release/deps/libthreadpool-433f648cab98e319.rlib" "/home/user/Projects/other/helix/target/release/deps/libnum_cpus-3c6425580b9bb41b.rlib" "/home/user/Projects/other/helix/target/release/deps/liblog-2ccdf85413674370.rlib" "/home/user/Projects/other/helix/target/release/deps/libtoml-9d655bc2734e9ee1.rlib" "/home/user/Projects/other/helix/target/release/deps/libetcetera-e5d618b204112366.rlib" "/home/user/Projects/other/helix/target/release/deps/libdirs_next-8dc8f94197ca47c5.rlib" "/home/user/Projects/other/helix/target/release/deps/libdirs_sys_next-c6b6fcb17840d3c7.rlib" "/home/user/Projects/other/helix/target/release/deps/liblibc-fda4f42ce649ce38.rlib" "/home/user/Projects/other/helix/target/release/deps/libthiserror-5b5148ffeaed0497.rlib" "/home/user/Projects/other/helix/target/release/deps/liblibloading-d1a1254a718c7133.rlib" "/home/user/Projects/other/helix/target/release/deps/libcfg_if-e9b9f12d2f61b1c5.rlib" "/home/user/Projects/other/helix/target/release/deps/libtree_sitter-a80c1fb2073a7d23.rlib" "/home/user/Projects/other/helix/target/release/deps/libregex-9a2fa22c47ca392d.rlib" "/home/user/Projects/other/helix/target/release/deps/libaho_corasick-647659da6883dadf.rlib" "/home/user/Projects/other/helix/target/release/deps/libmemchr-c2f809c40ec8e87a.rlib" "/home/user/Projects/other/helix/target/release/deps/libregex_syntax-04f6f521f41bf110.rlib" "/home/user/Projects/other/helix/target/release/deps/libserde-679d3a11164bb224.rlib" "/home/user/Projects/other/helix/target/release/deps/libanyhow-66aee07be193feca.rlib" "-Wl,--start-group" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-43e8dba7292d15f0.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libpanic_unwind-ef12f59e4a90f2f4.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libobject-677756f731266722.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libmemchr-87c7ca48d3b561b9.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libaddr2line-d59bbe3bfa8829d9.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libgimli-5027d692a3b9ae53.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_demangle-46a73cc9cde8d8ec.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd_detect-cadfc1a6cfd5f402.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libhashbrown-620a0879454632cb.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libminiz_oxide-47bb6ac5e94e5bb9.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libadler-0da6f0768ea742d3.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_alloc-4ec90f3d9851a42f.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libunwind-50a17a66e06c0d68.rlib" "-lunwind" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcfg_if-542d653a1361ebfa.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liblibc-7d1477e636eab66b.rlib" "-lc" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liballoc-c97029b44e59ba90.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_core-b32c2766aadb8b64.rlib" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcore-250003858c5ad774.rlib" "-Wl,--end-group" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-4d8fdcdce85aa3c2.rlib" "-Wl,-Bdynamic" "-ldl" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-nostartfiles" "-L" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib" "-L" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained" "-o" "/home/user/Projects/other/helix/target/release/build/helix-term-8116fac99c6ad7d6/build_script_build-8116fac99c6ad7d6" "-Wl,--gc-sections" "-static" "-no-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtend.o" "/home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtn.o"
  = note: /usr/lib/gcc/aarch64-alpine-linux-musl/11.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: /home/user/.rustup/toolchains/1.61.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-4d8fdcdce85aa3c2.rlib(cpu_model.o): in function `init_have_lse_atomics':
          /cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.70/./lib/builtins/cpu_model.c:811: undefined reference to `getauxval'
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
```

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -138,7 +138,7 @@ jobs:
 
       - name: Build AppImage
         shell: bash
-        if: matrix.build == 'x86_64-linux'
+        if: matrix.build == 'aarch64-linux' || matrix.build == 'x86_64-linux'
         run: |
           mkdir dist
 
@@ -147,8 +147,10 @@ jobs:
             name=${GITHUB_REF:10}
           fi
 
+          build="${{ matrix.build }}"
+
           export VERSION="$name"
-          export ARCH=x86_64
+          export ARCH=${build%-linux}
           export APP=helix
           export OUTPUT="helix-$VERSION-$ARCH.AppImage"
           export UPDATE_INFORMATION="gh-releases-zsync|$GITHUB_REPOSITORY_OWNER|helix|latest|$APP-*-$ARCH.AppImage.zsync"
@@ -245,7 +247,7 @@ jobs:
               mv bins-$platform/hx$exe $pkgname
               chmod +x $pkgname/hx$exe
 
-              if [[ "$platform" = "x86_64-linux" ]]; then
+              if [[ "$platform" = "aarch64-linux" || "$platform" = "x86_64-linux" ]]; then
                   mv bins-$platform/helix-*.AppImage* dist/
               fi
 

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -100,6 +100,14 @@ jobs:
           target: ${{ matrix.target }}
           override: true
 
+      # Install a pre-release version of Cross
+      # TODO: We need to pre-install Cross because we need cross-rs/cross#591 to
+      #       get a newer C++ compiler toolchain. Remove this step when Cross
+      #       0.3.0, which includes cross-rs/cross#591, is released.
+      - name: Install Cross
+        if: "matrix.cross"
+        run: cargo install cross --git https://github.com/cross-rs/cross.git --rev 47df5c76e7cba682823a0b6aa6d95c17b31ba63a
+
       - name: Run cargo test
         uses: actions-rs/cargo@v1
         if: "!matrix.skip_tests"

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -49,11 +49,11 @@ jobs:
           rust: stable
           target: x86_64-unknown-linux-gnu
           cross: false
-        # - build: aarch64-linux
-        #   os: ubuntu-20.04
-        #   rust: stable
-        #   target: aarch64-unknown-linux-gnu
-        #   cross: true
+        - build: aarch64-linux
+          os: ubuntu-20.04
+          rust: stable
+          target: aarch64-unknown-linux-gnu
+          cross: true
         - build: x86_64-macos
           os: macos-latest
           rust: stable

